src/gui/transferlistmodel.cpp

```diff
@@ -512,6 +531,8 @@ QVariant TransferListModel::internalValue(const BitTorrent::Torrent *torrent, co
         return QVariant::fromValue(torrent->infoHash().v2());
     case TR_REANNOUNCE:
         return torrent->nextAnnounce();
+    case TR_PRIVATE:
+        return privateSort(torrent->isPrivate(), torrent->hasMetadata());
```

Comment:
Why to complicate it?
```suggestion
        return (torrent->hasMetadata() ? torrent->isPrivate() : -1);
```

Reply:
Seems more correct would be the following:
```c++
return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());
```
...and use appropriate comparison in sort model (that takes into account invalid values).

Reply:
> Seems more correct would be the following:
> 
> ```c++
> return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());
> ```
> 
> ...and use appropriate comparison in sort model (that takes into account invalid values).

I implemented it:
https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistmodel.cpp#L528

Reply:
> I implemented it:
> 
> https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistmodel.cpp#L528

OK.
But you still haven't taken care of the second part:
>...and use appropriate comparison in sort model (that takes into account invalid values).

IIRC, the following will not handle invalid values properly:
```c++
threeWayCompare(leftValue.toInt(), rightValue.toInt());
```

Reply:
I am still not sure why we changed 
approach a)
> return (torrent->hasMetadata() ? torrent->isPrivate() : -1);

to approach b)
> return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());

With the first one, we always get an integer (1, 0, -1), since -1 is the fall-back and isPrivate() is only 1 or 0.
Since we convert the arguments for the threewaycompare explicitly to int, shouldn't it work if we sticked with approach a)?
https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistsortmodel.cpp#L204

Reply:
> I am still not sure why we changed approach a)
> 
> > return (torrent->hasMetadata() ? torrent->isPrivate() : -1);
> 
> to approach b)
> 
> > return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());

Because it better (more explicitly) reflects its essence. It either returns the value of "privateness "itself (of type bool), or invalid if it is not available (via an uninitialized QVariant). Another way is a workaround, since all values (both valid and not) are represented by another data type (integer). I.e. true/false/invalid vs 1/0/-1.

Reply:
> Since we convert the arguments for the threewaycompare explicitly to int

`internalValue()` should just return internal/raw representation of data. It shouldn't really depend on some particular usage of it.

Reply:
Thank you for the explanation, @glassez